### PR TITLE
Clarify the definition of AVIF_USE_CXX.

### PR DESCRIPTION
**Pull request by vrabaud**
Only set `AVIF_USE_CXX` when linking to C++ libraries: libyuv, libgav1 or ComplianceWarden 

This fixes https://github.com/AOMediaCodec/libavif/issues/2277

### CODE REVIEW

File: .github/workflows/ci-android-emulator-tests.yml [hunk 1]
```diff
@@ -21,20 +17,13 @@ concurrency:
 
 jobs:
   build-and-run-android-jni-tests:
-    runs-on: ubuntu-latest
+    runs-on: macos-latest
```
Comment (wantehchang) [inline .github/workflows/ci-android-emulator-tests.yml:20]: Undo the changes to this file. It looks like a merge conflict resolved incorrectly.

File: CMakeLists.txt [hunk 2]
```diff
@@ -632,13 +629,14 @@ endif()
...
     get_target_property(AOM_INTERFACE_LINK_LIBRARIES aom INTERFACE_LINK_LIBRARIES)
-    if(AOM_INTERFACE_LINK_LIBRARIES MATCHES vmaf)
-        # vmaf only affects the avifenc and avifdec targets below, not the avif library nor examples targets.
-        set(AVIF_USE_CXX ON)
-    endif()
 endif()
 
-if(AVIF_USE_CXX)
+if(AVIF_USE_CXX
+   OR AVIF_BUILD_APPS
```
Comment (wantehchang) [inline CMakeLists.txt:635]: Optional: Right now avifgainmaputil is the only app that uses C++, and it is an experimental feature. So this line can be changed to:

```
   OR (AVIF_BUILD_APPS AND AVIF_ENABLE_EXPERIMENTAL_GAIN_MAP)
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -133,17 +133,9 @@ option(
     OFF
 )
 
+# Whether the libavif library uses c++ indirectly (e.g. through linking to libgav1).
```
Comment (wantehchang) [inline CMakeLists.txt:136]: I suggest we rename the variable to reflect the new, more specific meaning. The new name should have "LIB" in the name, e.g., "AVIF_LIB_USE_CXX" or "AVIF_USE_CXX_IN _LIB" or "AVIF_LINK_LIB_AS_CXX".

File: CMakeLists.txt [hunk 4]
```diff
@@ -571,6 +565,20 @@ if(NOT AVIF_CODEC_AOM_ENABLED
...
+   OR (AVIF_CODEC_AOM_ENABLED AND (AOM_INTERFACE_LINK_LIBRARIES MATCHES vmaf))
+)
+    enable_language(CXX)
+    if(AVIF_ENABLE_NODISCARD)
+        set(CMAKE_CXX_STANDARD 17)
+    else()
+        set(CMAKE_CXX_STANDARD 14)
+    endif()
+endif()
+
```
Comment (wantehchang) [inline CMakeLists.txt:581]: Delete the if statement at lines 648-650 because the variable `AOM_INTERFACE_LINK_LIBRARIES` is unused now:

```
if(AVIF_CODEC_AOM_ENABLED)
    get_target_property(AOM_INTERFACE_LINK_LIBRARIES aom INTERFACE_LINK_LIBRARIES)
endif()
```

File: CMakeLists.txt [hunk 5]
```diff
@@ -571,6 +565,20 @@ if(NOT AVIF_CODEC_AOM_ENABLED
     message(WARNING "libavif: No decoding library is enabled.")
 endif()
 
+if(AVIF_USE_CXX
+   OR (AVIF_BUILD_APPS AND AVIF_ENABLE_EXPERIMENTAL_GAIN_MAP)
+   OR AVIF_ENABLE_FUZZTEST
+   OR AVIF_ENABLE_GTEST
+   OR (AVIF_CODEC_AOM_ENABLED AND (AOM_INTERFACE_LINK_LIBRARIES MATCHES vmaf))
```
Comment (wantehchang) [inline CMakeLists.txt:572]: IMPORTANT: With the new meaning of `AVIF_USE_CXX` defined at line 136, I would expect that this case falls under the new definition of `AVIF_USE_CXX` and therefore should be handled by setting `AVIF_USE_CXX` to `ON`.

However, the original comment has a comment (which may have been deleted accidentally in this pull request):

```
if(AVIF_CODEC_AOM_ENABLED)
    get_target_property(AOM_INTERFACE_LINK_LIBRARIES aom INTERFACE_LINK_LIBRARIES)
    if(AOM_INTERFACE_LINK_LIBRARIES MATCHES vmaf)
        # vmaf only affects the avifenc and avifdec targets below, not the avif library nor examples targets.
        set(AVIF_USE_CXX ON)
    endif()
endif()
```

If that comment is accurate, then we do NOT need to link the libavif library using the C++ compiler. So we need to look into this. If we are not sure if that comment is wrong, it is best to assume the comment is accurate and preserve the original behavior.

Comment (vrabaud) [reply CMakeLists.txt:578]: Bad copy paste: this will be fixed once https://github.com/AOMediaCodec/libavif/issues/2274 is fixed

File: CMakeLists.txt (hunk 3 above)
Comment (vrabaud) [reply CMakeLists.txt:136]: AVIF_LIB_USE_CXX ssems like the best one to me

**Review by wantehchang — APPROVED** [review]
LGTM. Thanks! I suggest some small changes.

**Comment by wantehchang** [comment]
I tested this pull request (both without and with my suggested changes) using the following commands:
```
cd ext
./aom.cmd
cd ..
mkdir build
cd build
cmake .. -G Ninja -DAVIF_CODEC_AOM=LOCAL -DAVIF_LIBYUV=OFF -DAVIF_BUILD_APPS=ON
ninja
```

Before:

```
[34/40] Linking CXX shared library libavif.so.16.1.0
...
[39/40] Linking CXX executable avifdec
[40/40] Linking CXX executable avifenc
```

After:

```
[34/40] Linking C shared library libavif.so.16.1.0
...
[39/40] Linking C executable avifdec
[40/40] Linking C executable avifenc
```

File: CMakeLists.txt (hunk 5 above)
Comment (wantehchang) [reply CMakeLists.txt:578]: I looked up the source of that comment. The comment was originally added in https://github.com/AOMediaCodec/libavif/commit/ef6da2a78 and updated in https://github.com/AOMediaCodec/libavif/commit/47e58af14. I believe the comment was added to describe the existing code. The code was originally added in https://github.com/AOMediaCodec/libavif/commit/778b77fa0. I think the code is incorrect. The comment accurately describes the code. That's why I find the comment confusing and seeminly incorrect.

This pull request corrects the code and deletes the comment (which becomes inaccurate because the code it described has changed). That's good.

File: CMakeLists.txt [hunk 6]
```diff
@@ -133,16 +133,8 @@ option(
...
-
-if(AVIF_BUILD_APPS
-   OR AVIF_ENABLE_FUZZTEST
-   OR AVIF_ENABLE_GTEST
-   OR AVIF_CODEC_LIBGAV1 STREQUAL "LOCAL"
-   OR AVIF_CODEC_LIBGAV1 STREQUAL "SYSTEM"
-)
-    set(AVIF_USE_CXX ON)
-endif()
+# Whether the libavif library uses c++ indirectly (e.g. through linking to libgav1).
```
[the remark this answers is not in the document]
Comment (vrabaud) [reply CMakeLists.txt:136]: done

File: CMakeLists.txt [hunk 7]
```diff
@@ -543,6 +537,12 @@ if(AVIF_CODEC_AOM_ENABLED)
 
     message(STATUS "libavif: Codec enabled: aom (${AVIF_CODEC_AOM_ENCODE_DECODE_CONFIG})")
     set(AVIF_PKG_CONFIG_EXTRA_REQUIRES_PRIVATE "${AVIF_PKG_CONFIG_EXTRA_REQUIRES_PRIVATE} aom")
+
+    get_target_property(AOM_INTERFACE_LINK_LIBRARIES aom INTERFACE_LINK_LIBRARIES)
+    if(AOM_INTERFACE_LINK_LIBRARIES MATCHES vmaf)
+        # vmaf only affects the avifenc and avifdec targets below, not the avif library nor examples targets.
```
[the remark this answers is not in the document]
Comment (vrabaud) [reply CMakeLists.txt:543]: done

File: CMakeLists.txt [hunk 8]
```diff
@@ -571,6 +571,20 @@ if(NOT AVIF_CODEC_AOM_ENABLED
     message(WARNING "libavif: No decoding library is enabled.")
 endif()
 
+if(AVIF_LIB_USE_CXX
+   OR (AVIF_BUILD_APPS AND AVIF_ENABLE_EXPERIMENTAL_GAIN_MAP)
+   OR AVIF_ENABLE_FUZZTEST
+   OR AVIF_ENABLE_GTEST
```
[the remark this answers is not in the document]
Comment (vrabaud) [reply CMakeLists.txt:577]: let's do it here.

**Review by wantehchang — COMMENTED** [review]
I reviewed the last (fifth) commit ("Fix nits"). LGTM. Thanks!
